### ISSUE TITLE
lsp sometimes turns into zombie process

### ISSUE THREAD

**Issue by tyler338** [denoland/vscode_deno#374]
**To Reproduce**
1. open vscode
2. play around in editor for like a minute
3. close vscode with cmd+q
4. lsp turns into zombie process

*note: doesn't always happen consistently, so it might require multiple attempts or different steps*

**Expected behavior**
lsp exits with vscode

**Screenshots**
![](https://cdn.discordapp.com/attachments/684898665151594506/822791887357345822/unknown.png)
**Versions**

vscode: 1.54.3 (Universal)
deno: 1.8.1+0e70d9e
extension: 3.2.0

**Comment by qwerasd205** [comment]
I have also experienced this issue (macOS)
It seems to occur most often (or perhaps most noticably) for me when I have large/complex javascript files open (For example, large obfuscated samples) which leaves the process using 100% of one of my CPU cores after quitting VS Code.

**Comment by imcotton** [comment]
Same here on **macOS**,  dangling `deno lsp` process each time open/close **vscode** while extension enabled within **Deno** project.

Right now I'm using following as temporary cleanup solution:

    pkill -f 'deno lsp'

**Comment by kidonng** [comment]
Just noticed this issue when browsing Activity Monitor. Literally freed 1GB worth of RAM after `killall deno`.

![image](https://user-images.githubusercontent.com/44045911/119254091-fb6c8f00-bbe6-11eb-9e40-a430244c5316.png)

There should be some kind of notice of this issue, or it is too subtle to to notice.

**Comment by KnorpelSenf** [comment]
Happens to me regularly when just _reloading_ VSCode (`Developer: Reload Window`). The extension even starts loads of zombie processes when I explicitly set `deno.enable: false` (before starting VSCode) in `.vscode/settings.json`. It's been this way for half a year or so, for multiple versions of both `deno` and this extension. Current versions are:

```bash
$ uname -a
Linux steffens-pc 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
$ deno --version 
deno 1.10.3 (release, x86_64-unknown-linux-gnu)
v8 9.1.269.27
typescript 4.2.2
$ code --version
1.56.2
054a9295330880ed74ceaedda236253b4f39a335
x64
```

**Comment by jeiea** [comment]
Same on windows.
```
> code --version
1.56.2
054a9295330880ed74ceaedda236253b4f39a335
x64
> deno --version
deno 1.10.3 (release, x86_64-pc-windows-msvc)
v8 9.1.269.27
typescript 4.2.2
```

**Comment by dsherret** [comment]
I investigated this and vscode-languageserver-node was sending an exit notification then closing its connection before the notification was sent. I opened https://github.com/microsoft/vscode-languageserver-node/pull/776

**Comment by dsherret** [comment]
The vscode-languageclient exit notification not being sent issue is fixed and will be in the next version greater than 7.0.0; however, I believe there is another issue in lspower where `Server.serve` never seems to return after receiving the exit notification. I'm not sure about the best way of fixing this in lspower, but I've opened https://github.com/silvanshade/lspower/issues/10 to get more clarification.

In the meantime, we will have a "parent process checker" solution implemented in the next version of Deno (versions greater than v1.11.2) that will cause orphaned processes to terminate themselves shortly after vscode exits.

**Comment by kidonng** [comment]
@dsherret Thanks for working on this issue! However with VSCode 1.58.1, deno 1.11.5 and plugin 3.7.0 this issue still occurs. Meanwhile I discovered that the `--parent-pid` flag added to `deno lsp` doesn't work:

```
~                                                                                                        ✘ 0|1
❯ deno --version
deno 1.11.5 (release, x86_64-apple-darwin)
v8 9.1.269.35
typescript 4.3.2

~                                                                                                             
❯ deno lsp --parent-pid=1
error: Found argument '--parent-pid' which wasn't expected, or isn't valid in this context

USAGE:
    deno lsp [OPTIONS]

For more information try --help
```

**Comment by dsherret** [comment]
@kidonng it was switched to use the LSP in https://github.com/denoland/deno/pull/11083 because the LSP provides the parent process id, so actually this should have been fixed in v1.11.3 despite the problems in the current released version of vscode-languageclient and lspower.

Are you sure the issue still exists? I'm unable to reproduce. You might have to wait up to 30 seconds before the deno process terminates itself after vscode exits. Once vscode-languageclient does a new release and the bug in lspower is fixed and released then it should be instant.

**Comment by kidonng** [comment]
Yup, here is the evidence:
<img width="973" alt="Screen Shot 2021-07-31 at 23 08 29" src="https://user-images.githubusercontent.com/44045911/127744207-6a490fef-2522-4154-89f9-95f7a6d6da78.png">

**Comment by web2033** [comment]
The same behavior as 2 months ago when I first encountered it.
All latest versions.

**Comment by dsherret** [comment]
I just looked into this and I'm able to reproduce on Mac with the following steps (everything seems ok on linux):

1. Open vscode to a deno project.
2. Inspect the messages passed on the LSP and kill the parent PID provided in order to kill vscode.

What is very strange is that the parent process checker code does detect that the process is lost and it seems to call `std::process:exit(1)`, but the process never exits... I verified this by creating a dummy file right before the `std::process:exit(1)` call in https://github.com/denoland/deno/blob/8f00b5542caffd14988b923efe4f066b712d2858/cli/lsp/parent_process_checker.rs#L16

**Comment by arafatamim** [comment]
I can confirm the issue is present in Linux too. Every instance of a Deno project opened within VSCode  after exiting leaves the respective `deno lsp` process running even after some time has passed. (Deno 1.12.2, Linux 5.10.42-1-MANJARO)

`$ ps ux | grep deno`
```
tamim       3687  0.0  2.1 5321092 171156 ?      Sl   Jul31   0:09 deno lsp
tamim      55300  0.4  1.4 5184596 113736 ?      Sl   00:14   0:01 deno lsp
tamim      57471  0.0  0.0   8944  2424 pts/3    S+   00:21   0:00 grep deno
```

**Comment by dsherret** [comment]
Ok, thank you everyone. I found the issue and opened https://github.com/denoland/deno/pull/11562

**Comment by web2033** [comment]
For Windows too?

**Comment by dsherret** [comment]
This should be resolved now in Deno 1.13. Please let me know if you still experience any issues.

The process should terminate itself within 30 seconds of vscode closing. For the process to exit immediately after vscode close, we are still waiting on a new release of vscode-languageclient and for the bug to be fixed in lspower.

@web2033 yes. Should be fixed for all platforms.
